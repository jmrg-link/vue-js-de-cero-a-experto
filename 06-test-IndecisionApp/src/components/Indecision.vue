<template>
  <div>
    <h1>IndecisionApp</h1>
    <img v-if="img"  :src="img" alt="bg" />
    <div class="bg-dark"></div>
    <div class="indecision-container">
      <input v-model="question" type="text" placeholder="Realiza una pregunta..." />
      <p>Recuerda terminar con un signo de interrogaci√≥n <strong>(?)</strong></p>
      <div v-if="isValidQuestion">
        <h2>{{question}}</h2>
        <h1>{{answer}}</h1>
        <!-- <h1>{{answer === 'yes' ? 'Sii!' : 'Noo!'}}</h1> -->
      </div>
    </div>
  </div>
</template>
<script>
export default {
  name: "IndecisionApp",
  data() {
      return {
          question: null,
          answer:null,
          img:null,
          isValidQuestion:false
      }
  },
  methods: {
      async getAnswer(){
          this.answer = 'pensando...'
          const {answer, image} = await fetch('https://yesno.wtf/api').then(response => response.json())
          this.answer = answer === 'yes' ? 'Sii!' : 'Noo!'
          this.img  = image
      }
  },
  watch:{
      question(value){
          this.isValidQuestion = false
          if(!value.includes('?')) return 
          this.isValidQuestion = true
          this.getAnswer()
      }
  }
}
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 5px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>