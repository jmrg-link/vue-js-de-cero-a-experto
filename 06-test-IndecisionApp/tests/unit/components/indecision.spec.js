import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision.vue";

describe("Indecision component", () => {
  let wrapper;
  let clgSpy;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    clgSpy = jest.spyOn(console, "log");
    jest.clearAllMocks();
  });

  test("1ª Debe de hacer match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("2ª Escribir en el input no debe disparar nada ", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");
    await input.setValue("Hola mundo");
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test('3ª Esribir el simbolo de "?" debe de disparar el getAnswer ', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");
    await input.setValue("test?");
    expect(clgSpy).toHaveBeenCalledTimes(2);
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  test("4ª Pruebas en getAnswer ", async () => {
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("Sii!");
  });

  test("5ª Pruebas en getAnswer - Fallo de la API ", async () => {
    //TODO: fallo en el api
    fetch.mockImplementationOnce(() => Promise.reject("API is down "));
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("No se pudo mostrar la API");
  });
});
