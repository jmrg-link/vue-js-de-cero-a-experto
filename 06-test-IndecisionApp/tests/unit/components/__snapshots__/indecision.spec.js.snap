// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component 1ª Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>IndecisionApp</h1>
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Realiza una pregunta...">
    <p> Recuerda terminar con un signo de interrogación <strong>(?)</strong></p>
    <!--v-if-->
  </div>
</div>
`;
